File: app/backend/api/library.py
"""Library CRUD endpoints."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query

from models.schemas import LibraryListResponse, TrackDetailResponse, TrackUpdateRequest
from services import library_service

router = APIRouter(prefix="/api/library", tags=["library"])


@router.get("", response_model=LibraryListResponse)
async def list_tracks(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    search: str = "",
    sort: str = "created_at",
    order: str = "desc",
):
    """List tracks with pagination and search."""
    return await library_service.list_tracks(
        page=page, page_size=page_size, search=search, sort=sort, order=order
    )


@router.get("/{track_id}", response_model=TrackDetailResponse)
async def get_track(track_id: str):
    """Get track details including stems."""
    result = await library_service.get_track(track_id)
    if not result:
        raise HTTPException(404, "Track not found")
    return result


@router.patch("/{track_id}")
async def update_track(track_id: str, request: TrackUpdateRequest):
    """Update track metadata (title, tags)."""
    ok = await library_service.update_track(
        track_id, title=request.title, tags=request.tags
    )
    if not ok:
        raise HTTPException(404, "Track not found")
    return {"status": "ok"}


@router.delete("/{track_id}")
async def delete_track(track_id: str):
    """Delete a track and its files."""
    ok = await library_service.delete_track(track_id)
    if not ok:
        raise HTTPException(404, "Track not found")
    return {"status": "ok"}

File: app/frontend/src/api/library.ts
import { api } from './client'
import type { LibraryListResponse, TrackDetailResponse } from '../types/api'

/**
 * List tracks with pagination, search, sorting, and optional favorite filter.
 */
export async function listTracks(
  page?: number,
  search?: string,
  sort?: string,
  order?: string,
  favorite?: boolean,
): Promise<LibraryListResponse> {
  const params = new URLSearchParams()
  if (page !== undefined) params.set('page', String(page))
  if (search) params.set('search', search)
  if (sort) params.set('sort', sort)
  if (order) params.set('order', order)
  if (favorite !== undefined) params.set('favorite', String(favorite))

  const query = params.toString()
  return api.get<LibraryListResponse>(`/library${query ? `?${query}` : ''}`)
}

/**
 * Get track details including stems.
 */
export async function getTrack(id: string): Promise<TrackDetailResponse> {
  return api.get<TrackDetailResponse>(`/library/${id}`)
}

/**
 * Update track metadata (title, tags, favorite, rating).
 */
export async function updateTrack(
  id: string,
  data: {
    title?: string
    tags?: string
    favorite?: boolean
    rating?: number
  },
): Promise<void> {
  await api.patch(`/library/${id}`, data)
}

/**
 * Delete a track and its files.
 */
export async function deleteTrack(id: string): Promise<void> {
  await api.delete(`/library/${id}`)
}
